modules/js-and-css/speculation-rules/settings.php

```diff
@@ -0,0 +1,218 @@
+ *
+ * @since n.e.x.t
+ *
+ * @return array Associative array of `$eagerness => $label` pairs.
+ */
+function plsr_get_eagerness_labels() {
+	return array(
+		'conservative' => _x( 'Conservative (typically on click)', 'setting label', 'performance-lab' ),
+		'moderate'     => _x( 'Moderate (typically on hover)', 'setting label', 'performance-lab' ),
+		'eager'        => _x( 'Eager (on slightest suggestion)', 'setting label', 'performance-lab' ),
+	);
+}
+
+/**
+ * Returns the default setting value for Speculation Rules configuration.
+ *
+ * @since n.e.x.t
+ *
+ * @return array Default value, an associative array with 'mode' and 'eagerness' keys.
+ */
+function plsr_get_setting_default() {
+	return array(
+		'mode'      => 'prerender',
+		'eagerness' => 'moderate',
+	);
+}
+
+/**
+ * Sanitizes the setting for Speculation Rules configuration.
+ *
+ * @since n.e.x.t
+ *
+ * @param mixed $input Setting to sanitize.
+ * @return array Sanitized setting, an associative array with 'mode' and 'eagerness' keys.
+ */
+function plsr_sanitize_setting( $input ) {
+	$default_value = plsr_get_setting_default();
+
+	if ( ! is_array( $input ) ) {
+		return $default_value;
+	}
+
+	$mode_labels      = plsr_get_mode_labels();
+	$eagerness_labels = plsr_get_eagerness_labels();
+
+	// Ensure only valid keys are present.
+	$value = array_intersect_key( $input, $default_value );
+
+	// Set any missing or invalid values to their defaults.
+	if ( ! isset( $value['mode'] ) || ! isset( $mode_labels[ $value['mode'] ] ) ) {
+		$value['mode'] = $default_value['mode'];
+	}
+	if ( ! isset( $value['eagerness'] ) || ! isset( $eagerness_labels[ $value['eagerness'] ] ) ) {
+		$value['eagerness'] = $default_value['eagerness'];
+	}
+
+	return $value;
+}
+
+/**
+ * Registers setting to control Speculation Rules configuration.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+function plsr_register_setting() {
+	register_setting(
+		'reading',
+		'plsr_speculation_rules',
+		array(
+			'type'              => 'object',
+			'description'       => __( 'Configuration for the Speculation Rules API.', 'performance-lab' ),
+			'sanitize_callback' => 'plsr_sanitize_setting',
+			'default'           => plsr_get_setting_default(),
+			'show_in_rest'      => array(
+				'schema' => array(
+					'properties' => array(
+						'mode'      => array(
+							'description' => __( 'Whether to prefetch or prerender URLs.', 'performance-lab' ),
+							'type'        => 'string',
+							'enum'        => array_keys( plsr_get_mode_labels() ),
+						),
+						'eagerness' => array(
+							'description' => __( 'Whether to trigger on click (conservative), on hover (moderate), or on even the slight suggestion (eager) that the user may navigate to the URL.', 'performance-lab' ),
+							'type'        => 'string',
+							'enum'        => array_keys( plsr_get_eagerness_labels() ),
+						),
+					),
+				),
+			),
+		)
+	);
+}
+add_action( 'init', 'plsr_register_setting' );
```

Comment:
Shoudn't this this rather be at `admin_init` and `rest_api_init`?

```suggestion
add_action( 'admin_init', 'plsr_register_setting' );
add_action( 'rest_api_init', 'plsr_register_setting' );
```

Reply:
Options in core are registered on `init` I believe, which I think is right because options can in principle be set in any use-case, not just limited to WP Admin and REST API.

Reply:
But this is registering the setting, not the option, right? AFAIK, the setting only needs to be registered in a context where it can be modified. I'm only seeing `register_setting()` being [added](https://github.com/WordPress/wordpress-develop/blob/35add7378e093c8808ef9c08ee4b07f71b14963d/src/wp-includes/blocks/site-logo.php#L61-L95) at `rest_api_init` and Twenty Fourteen [registers](https://github.com/WordPress/wordpress-develop/blob/35add7378e093c8808ef9c08ee4b07f71b14963d/src/wp-content/themes/twentyfourteen/inc/featured-content.php#L82) a setting at `admin_init`.

Reply:
> But this is registering the setting, not the option, right?

Setting and option are pretty much the same thing. There will never be a `register_option()` as `register_setting()` covers everything that it could possibly do. There are historic reasons for the weird naming. And FWIW some core contributors have previously said in discussions that they wish `register_setting()` was called `register_option()`.

> AFAIK, the setting only needs to be registered in a context where it can be modified. I'm only seeing `register_setting()` being [added](https://github.com/WordPress/wordpress-develop/blob/35add7378e093c8808ef9c08ee4b07f71b14963d/src/wp-includes/blocks/site-logo.php#L61-L95) at `rest_api_init` and Twenty Fourteen [registers](https://github.com/WordPress/wordpress-develop/blob/35add7378e093c8808ef9c08ee4b07f71b14963d/src/wp-content/themes/twentyfourteen/inc/featured-content.php#L82) a setting at `admin_init`.

Settings can be modified anywhere. Where you register a setting can depend on where you want to have its registered behavior supported, but it doesn't change where a setting can be modified. For example, the `default` parameter that can be registered is relevant literally anywhere that `get_option()` is called, so this is a good example of why _general_ registration (on `init`) makes most sense. A lot of the confusion is most likely for historic reasons:
* Many years ago, before the REST API, the only purpose of registering settings was to use the Settings API, which is the specific API to render UI for options in WP Admin (for whatever reason the "admin options API" was called the "Settings API"). That's why it only was relevant on `admin_init`.
* When the REST API came, registering settings was expanded to do a lot more than just registering for that admin Settings API. The decision was made to build on top of the existing function rather than introducing a more generic `register_option()` function. Myself and others don't necessarily think that was the right call, but we have to live with it now. Initially, some code then started to also add setting registration to `rest_api_init` when it was relevant there.
* However, both of those are just two specific use-cases for settings. There is nothing in WordPress that ties setting options to those contexts though. For the most reliable behavior, settings need to be registered for use anywhere (unless it's legacy usage that only registers specifically for the old school Settings API, i.e. without all the new stuff like `type`, `description` etc.). Otherwise you run into weird situations where `get_option()` sometimes returns a useful default by default and other times not, which I'm not sure is useful in any case.

Reply:
This is one of my (least?) favorite annoyances and weirdnesses in WordPress, so I just thought I'd start a poll: https://twitter.com/felixarntz/status/1747349683947880931

Reply:
Init is also important for things like WP-CLI, FWIW